README.md
<p align="center">
<img src="https://d3co7cvuqq9u2k.cloudfront.net/public/image/logo/uiza_logo_color.png" data-canonical-src="https://uiza.io" width="450" height="220" />
</p>

# Uiza Player Web SDK

[![License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://npmjs.org/package/@uizaio/playerjs)
[![NPM version](http://img.shields.io/npm/dm/@uizaio/playerjs.svg?style=flat)](https://npmjs.org/package/@uizaio/playerjs)
[![Build Status](https://travis-ci.org/uizaio/uiza-javascript-sdk-player.svg?branch=master)](https://travis-ci.org/uizaio/uiza-javascript-sdk-player)
[![dependencies Status](https://david-dm.org/uizaio/uiza-javascript-sdk-player/status.svg)](https://david-dm.org/uizaio/uiza-javascript-sdk-player)
[![devDependencies Status](https://david-dm.org/uizaio/uiza-javascript-sdk-player/dev-status.svg)](https://david-dm.org/uizaio/uiza-javascript-sdk-player?type=dev)
[![](https://data.jsdelivr.com/v1/package/npm/@uizaio/playerjs/badge)](https://www.jsdelivr.com/package/npm/@uizaio/playerjs)

[![NPM](https://nodei.co/npm/@uizaio/playerjs.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/@uizaio/playerjs/)

Uiza Player is a simple, lightweight, accessible and customizable HTML5, media player that supports [_modern_](#browser-support) browsers.


### Demos

You can try Uiza in Codepen using our minimal templates. For Streaming we also have example integrations with: [Hls.js](https://playcode.io/614075), [Dash.js](https://playcode.io/628492) and [VOD](https://playcode.io/614073/)

# Quick setup

## HTML

Uiza Player extends upon the standard [HTML5 media element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement) markup so that's all you need for those types.

| Attribute | Value |Description |
| -- | --  | -- |
| `autoplay` | autoplay | Specifies that the video will start playing as soon as it is ready |
| `controls` | controls | Specifies that video controls should be displayed (such as a play/pause button etc). |
| `src` | URL | Specifies the URL of the video file |
| `height` | pixels | Sets the height of the video player |
| `width` | pixels | Sets the width of the video player |
| `loop` | loop | Specifies that the video will start over again, every time it is finished |
| `muted` | muted | Specifies that the audio output of the video should be muted |
| `poster` | URL | Specifies an image to be shown while the video is downloading, or until the user hits the play button |
| `preload` | auto / metadata / none | Specifies if and how the author thinks the video should be loaded when the page loads |


### HLS Stream Video

```html
<video id='playerUz' crossorigin playsinline autoplay></video>
<script>
  var videoStreamURL = 'https://1955897154.rsc.cdn77.org/live/116881ad-3463-4209-b16f-50e0e48add10/master.m3u8?cm=eyJlbnRpdHlfaWQiOiI2ZWUyMzk5NS1mNGFhLTQ1ZDMtOTM5NS0xN2NlNjBlZWJjNDAiLCJlbnRpdHlfc291cmNlIjoibGl2ZSIsImFwcF9pZCI6IjkyYThkMjAzMmZlODQ5MmFhNzc4MDRiNGMyYzUxOWM1In0=';

(async (options) => {
  var uiza = await new Uiza(document.getElementById('playerUz'), options);
  return { uiza, time: Date.now() };
})({src: videoStreamURL});
</script>
```

See [initialising](#initialising) for more information on advanced setups.

You can use our CDN for the JavaScript. There's 2 versions; one with and one without [polyfills](#polyfills). My recommendation would be to manage polyfills seperately as part of your application but to make life easier you can use the polyfilled build.

```html
<script src="https://sdk.uiza.io/v5/release/uiza.min.js"></script>
```

...or...

```html
<script src="https://sdk.uiza.io/v5/release/uiza.polyfilled.js"></script>
```

## CSS

Include the `uiza.css` stylsheet into your `<head>`.

```html
<link rel="stylesheet" href="https://sdk.uiza.io/v5/release/uiza.css" />
```

## Example

example.html

```html
<!DOCTYPE html>
<html lang='en' dir='ltr'>

<head>
  <meta charset='utf-8'>
  <title></title>
  <meta name='viewport' content='width=device-width,initial-scale=1'>

  <link href='https://sdk.uiza.io/v5/release/uiza.css' rel='stylesheet'>
  <script src='https://sdk.uiza.io/v5/release/uiza.min.js'></script>

</head>

<body>
  <video id='playerUz' crossorigin playsinline autoplay></video>

  <script>
    var videoStreamURL = 'https://1955897154.rsc.cdn77.org/live/116881ad-3463-4209-b16f-50e0e48add10/master.m3u8?cm=eyJlbnRpdHlfaWQiOiI2ZWUyMzk5NS1mNGFhLTQ1ZDMtOTM5NS0xN2NlNjBlZWJjNDAiLCJlbnRpdHlfc291cmNlIjoibGl2ZSIsImFwcF9pZCI6IjkyYThkMjAzMmZlODQ5MmFhNzc4MDRiNGMyYzUxOWM1In0=';

    ((options) => {
      window.UZplayer = new Uiza(document.getElementById('playerUz'), options);
    })({ src: videoStreamURL });
  </script>

</body>

</html>
```


# Advanced

## Customizing HLS buffer

```
buffer: {
  maxBufferLength: 15
  maxMaxBufferLength: 90
}
```

`maxBufferLength` (default: 10 seconds, maximum 30 seconds)

Maximum buffer length in seconds. If buffer length is/become less than this value, a new fragment will be loaded. This is the guaranteed buffer length hls.js will try to reach, regardless of maxBufferSize.

`maxMaxBufferLength`: (default 60 seconds, maximum 120 seconds)
Maximum buffer length in seconds. Hls.js will never exceed this value, even if maxBufferSize is not reached yet.

hls.js tries to buffer up to a maximum number of bytes (6 MB by default) rather than to buffer up to a maximum nb of seconds. this is to mimic the browser behaviour (the buffer eviction algorithm is starting after the browser detects that video buffer size reaches a limit in bytes)

Example for maxBufferLengthand maxMaxBufferLength.  [Demo here](https://playcode.io/628085)
```html
<!DOCTYPE html>
<html lang='en' dir='ltr'>

<head>
  <meta charset='utf-8'>
  <title></title>
  <meta name='viewport' content='width=device-width,initial-scale=1'>

  <link href='https://sdk.uiza.io/v5/release/uiza.css' rel='stylesheet'>
  <script src='https://sdk.uiza.io/v5/release/uiza.min.js'></script>

</head>

<body>
  <video id='playerUz' crossorigin playsinline autoplay></video>

  <script>
    var videoStreamURL = 'https://1955897154.rsc.cdn77.org/live/116881ad-3463-4209-b16f-50e0e48add10/master.m3u8?cm=eyJlbnRpdHlfaWQiOiI2ZWUyMzk5NS1mNGFhLTQ1ZDMtOTM5NS0xN2NlNjBlZWJjNDAiLCJlbnRpdHlfc291cmNlIjoibGl2ZSIsImFwcF9pZCI6IjkyYThkMjAzMmZlODQ5MmFhNzc4MDRiNGMyYzUxOWM1In0=';

    ((options) => {
      window.UZplayer = new Uiza(document.getElementById('playerUz'), options);
    })({
      buffer: {
        maxBufferLength: 10,
        maxMaxBufferLength: 60
      },
      src: videoStreamURL
    });
  </script>

</body>

</html>
```

## Customizing the UI button

| Attribute | Default | Description |
| -- | --  | -- |
| `restart` | `true` | Only for VOD, Set the value to `false` to disable restart button on the video player |
| `rewind` | `true` | Only for VOD, Set the value to `false` to disable rewind button on the video player |
| `play` | `true` | Set the value to `false` to disable play button on the video player. Remember: should autoplay on video tag |
| `mute` | `true` | Set the value to `false` to disable mute button on the video player |
| `volume` | `true` | Set the value to `false` to disable volume button on the video player |
| `currentTime` | `true` | Set the value to `false` to hide the current time on the video player |
| `toggleLiveViewer` | `true` | Set the value to `false` to hide the viewer count on the video player |
| `progress` | `true` | Set the value to `false` to hide the progress bar on the video player |
| `fastForward` | `true` | Set the value to `false` to disable the fast forward button on the video player |
| `captions` | `true` | Set the value to `false` to disable the captions button on the video player |
| `settings` | `true` | Set the value to `false` to disable the settings button on the video player |
| `pip` | `true` | Set the value to `false` to disable the picture-in-picture button on the video player |
| `live` | `true` | Set the value to `false` to hide the live button on the video player |
| `fullscreen` | `true` | Set the value to `false` to disable the fullscreen button on the video player |
| `speed` | `true` | Set the value to `false` to disable the speed button on the video player |
```
ui: {
    restart: true,
    rewind: true,
    play: true,
    mute: true,
    volume: true,
    currentTime: true,
    toggleLiveViewer: true,
    fastForward: true,
    progress: true,
    captions: true,
    settings: true,
    pip: true,
    live: true,
    fullscreen: true,
    speed: true
}
```

Example for hide button `pip` and `settings`, only override configure to false. [Demo here](https://playcode.io/608599/)
```html
<!DOCTYPE html>
<html lang='en' dir='ltr'>

<head>
  <meta charset='utf-8'>
  <title></title>
  <meta name='viewport' content='width=device-width,initial-scale=1'>

  <link href='https://sdk.uiza.io/v5/release/uiza.css' rel='stylesheet'>
  <script src='https://sdk.uiza.io/v5/release/uiza.min.js'></script>

</head>

<body>
  <video id='playerUz' crossorigin playsinline autoplay muted></video>

  <script>
    var videoStreamURL = 'https://1955897154.rsc.cdn77.org/live/116881ad-3463-4209-b16f-50e0e48add10/master.m3u8?cm=eyJlbnRpdHlfaWQiOiI2ZWUyMzk5NS1mNGFhLTQ1ZDMtOTM5NS0xN2NlNjBlZWJjNDAiLCJlbnRpdHlfc291cmNlIjoibGl2ZSIsImFwcF9pZCI6IjkyYThkMjAzMmZlODQ5MmFhNzc4MDRiNGMyYzUxOWM1In0=';

    ((options) => {
      window.UZplayer = new Uiza(document.getElementById('playerUz'), options);
    })({
      ui: {
        pip: false,
        settings: false
      }, 
      src: videoStreamURL
    });
  </script>

</body>

</html>
```

## Customizing the CSS

If you want to change any design tokens used for the rendering of the player, you can do so using [CSS Custom Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).

Here's a list of the properties and what they are used for:

| Name | Description | Default / Fallback |
| -- | -- | -- |
| `--uiza-color-main` | The primary UI color. | ![#f03c15](https://placehold.it/15/00b3ff/000000?text=+) `#00b3ff` |
| `--uiza-tab-focus-color` | The color used for the dotted outline when an element is `:focus-visible` (equivalent) keyboard focus.  | `--uiza-color-main` |
| `--uiza-badge-background` | The background color for badges in the menu. | ![#4a5464](https://placehold.it/15/4a5464/000000?text=+) `#4a5464` | 
| `--uiza-badge-text-color` | The text color for badges. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-badge-border-radius` | The border radius used for badges. | `2px` |
| `--uiza-tab-focus-color` | The color used to highlight tab (keyboard) focus.                                                       | `--uiza-color-main` |
| `--uiza-captions-background` | The color for the background of captions. | `rgba(0, 0, 0, 0.8)` |
| `--uiza-captions-text-color` | The color used for the captions text. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-control-icon-size` | The size of the icons used in the controls. | `18px` |
| `--uiza-control-spacing` | The space between controls (sometimes used in a multiple - e.g. `10px / 2 = 5px`).                      | `10px` |
| `--uiza-control-padding` | The padding inside controls.                                                                            | `--uiza-control-spacing * 0.7` (`7px`) |
| `--uiza-control-radius` | The border radius used on controls. | `3px` |
| `--uiza-control-toggle-checked-background` | The background color used for checked menu items. | `--uiza-color-main` |
| `--uiza-video-controls-background` | The background for the video controls. | `linear-gradient(rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.75))` |
| `--uiza-video-control-color` | The text/icon color for video controls. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-video-control-color-hover` | The text/icon color used when video controls are `:hover`, `:focus` and `:focus-visible` (equivalent).  | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff`    |
| `--uiza-video-control-background-hover` | The background color used when video controls are `:hover`, `:focus` and `:focus-visible` (equivalent). | `--uiza-color-main`                                                   |
| `--uiza-audio-controls-background` | The background for the audio controls. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-audio-control-color` | The text/icon color for audio controls. | ![#4a5464](https://placehold.it/15/4a5464/000000?text=+) `#4a5464` |
| `--uiza-audio-control-color-hover` | The text/icon color used when audio controls are `:hover`, `:focus` and `:focus-visible` (equivalent).  | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-audio-control-background-hover`        | The background color used when video controls are `:hover`, `:focus` and `:focus-visible` (equivalent). | `--uiza-color-main`                                                   |
| `--uiza-menu-background` | The background color for menus. | `rgba(255, 255, 255, 0.9)` |
| `--uiza-menu-color` | The text/icon color for menu items. | ![#4a5464](https://placehold.it/15/4a5464/000000?text=+) `#4a5464` |
| `--uiza-menu-shadow` | The shadow used on menus. | `0 1px 2px rgba(0, 0, 0, 0.15)` |
| `--uiza-menu-radius` | The border radius on the menu. | `4px` |
| `--uiza-menu-arrow-size` | The size of the arrow on the bottom of the menu. | `6px` |
| `--uiza-menu-item-arrow-color` | The color of the arrows in the menu. | ![#728197](https://placehold.it/15/728197/000000?text=+) `#728197`    |
| `--uiza-menu-item-arrow-size` | The size of the arrows in the menu. | `4px` |
| `--uiza-menu-border-color` | The border color for the bottom of the back button in the top of the sub menu pages. | ![#dcdfe5](https://placehold.it/15/dcdfe5/000000?text=+) `#dcdfe5` |
| `--uiza-menu-border-shadow-color` | The shadow below the border of the back button in the top of the sub menu pages. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff`    |
| `--uiza-progress-loading-size` | The size of the stripes in the loading state in the scrubber. | `25px` |
| `--uiza-progress-loading-background` | The background color on the loading state in the scrubber. | `rgba(35, 40, 47, 0.6)` |
| `--uiza-video-progress-buffered-background` | The fill color for the buffer indication in the scrubber for video. | `rgba(255, 255, 255, 0.25)` |
| `--uiza-audio-progress-buffered-background` | The fill color for the buffer indication in the scrubber for audio. | `rgba(193, 200, 209, 0.6)` |
| `--uiza-range-thumb-height` | The height of the scrubber handle/thumb. | `13px` |
| `--uiza-range-thumb-background` | The background of the scrubber handle/thumb. | ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `#ffffff` |
| `--uiza-range-thumb-shadow` | The shadow of the scrubber handle/thumb. | `0 1px 1px rgba(215, 26, 18, 0.15), 0 0 0 1px rgba(215, 26, 18, 0.2)` |
| `--uiza-range-thumb-active-shadow-width` | The width of the shadow when the scrubber handle/thumb is `:active` (pressed). | `3px` |
| `--uiza-range-track-height` | The height of the scrubber/progress track. | `5px` |
| `--uiza-range-fill-background` | The fill color of the scrubber/progress. | `--uiza-color-main` |
| `--uiza-video-range-track-background` | The background of the scrubber/progress. | `--uiza-video-progress-buffered-background` |
| `--uiza-video-range-thumb-active-shadow-color` | The color of the shadow when the video scrubber handle/thumb is `:active` (pressed). | `rgba(255, 255, 255, 0.5)` |
| `--uiza-audio-range-track-background` | The background of the scrubber/progress. | `--uiza-video-progress-buffered-background` |
| `--uiza-audio-range-thumb-active-shadow-color` | The color of the shadow when the audio scrubber handle/thumb is `:active` (pressed). | `rgba(215, 26, 18, 0.1)` |
| `--uiza-tooltip-background` | The background color for tooltips. | `rgba(255, 255, 255, 0.9)` |
| `--uiza-tooltip-color` | The text color for tooltips. | ![#4a5464](https://placehold.it/15/4a5464/000000?text=+) `#4a5464` |
| `--uiza-tooltip-padding` | The padding for tooltips. | `calc(var(--uiza-control-spacing) / 2))` |
| `--uiza-tooltip-arrow-size` | The size of the arrow under tooltips. | `4px` |
| `--uiza-tooltip-radius` | The border radius on tooltips. | `3px` |
| `--uiza-tooltip-shadow` | The shadow on tooltips. | `0 1px 2px rgba(0, 0, 0, 0.15)` |
| `--uiza-font-family` | The font family used in the player. | |
| `--uiza-font-size-base` | The base font size. Mainly used for captions. | `15px` |
| `--uiza-font-size-small` | The smaller font size. Mainly used for captions. | `13px` |
| `--uiza-font-size-large` | The larger font size. Mainly used for captions. | `18px` |
| `--uiza-font-size-xlarge` | The even larger font size. Mainly used for captions. | `21px` |
| `--uiza-font-size-time` | The font size for the time. | `--uiza-font-size-small` |
| `--uiza-font-size-menu` | The font size used in the menu. | `--uiza-font-size-small` |
| `--uiza-font-size-badge` | The font size used for badges. | `9px` |
| `--uiza-font-weight-regular` | The regular font weight. | `400` |
| `--uiza-font-weight-bold` | The bold font weight. | `600` |
| `--uiza-line-height` | The line height used within the player. | `1.7` |
| `--uiza-font-smoothing` | Whether to enable font antialiasing within the player. | `false` |

You can set them in your CSS for all players:

```css
:root {
  --uiza-color-main: #1ac266;
}
```

...or for a specific class name:

```css
.player {
  --uiza-color-main: #1ac266;
}
```

...or in your HTML:

```html
<video class="player" style="--uiza-color-main: #1ac266;">
    ...
</vieo>
```

## Cross Origin (CORS)


CHANGELOG.md
[comment]: <> (Follow this to sort the changelog items:)
[comment]: <> (+ Feature block, sort by a, b, c,...)
[comment]: <> (+ Bugs fixed block, sort by a, b, c,...)
[comment]: <> (+ Demo page, sort by a, b, c,...)

## v5.20.7

- Support play with [DashJS](https://github.com/Dash-Industry-Forum/dash.js/wiki)
- Auto provider by URL of the video
- Lazyload library for dashjs or hlsjs
- Fixbug change quality with hlsjs
- Play default with non-timeshift by the hls.js link from Uiza (`#EXT-X-UZ-TIMESHIFT:extras/master.m3u8` or `#EXT-X-UZ-TIMESHIFT:master.m3u8`)


## v5.20.06rc6

- Add context menu on right click
- Add [toggle UI](https://github.com/uizaio/uiza-javascript-sdk-player/blob/master/README.md#customizing-the-ui-button)
- Add plugin HLS stats
- Add plugin timeshift for HLS livestream
- Add plugin Analytics for show viewer/live button
- Add plugin Embed and Share


## v5.0.0-alpha.20200102 (2020-01-02) - Local release

- Fix: playback be URL encoded


## v5.0.0-alpha.20191227 (2019-12-27) - Local release

- Load static from cloud
- Config: retry
- Support iframe


## v5.0.0-alpha.20191111 (2019-11-11) - Local release

- Debugger mode
- Keyboard shortcut
- Optimize startup time
- Picture in picture
- Preview & thumbnails
- Remove focus border
- Responsive
- Setting: captions
- Demo: captions, preview & thumbnails, debugger


## v5.0.0-alpha.20191022 (2019-10-22) - Local release

- Default skin
- HTML5 playback
- HLS playback
- Playback API
- Settings: speed, quality
- Demo: VOD & live


src/js/uiza.d.ts
// Type definitions for uiza v5
// Project: https://uiza.io
// TypeScript Version: 3.0

export = Uiza;
export as namespace Uiza;

declare class Uiza {
  /**
   * Setup a new instance
   */
  static setup(targets: NodeList | HTMLElement | HTMLElement[] | string, options?: Uiza.Options): Uiza[];

  /**
   * Check for support
   * @param mediaType
   * @param provider
   * @param playsInline Whether the player has the playsinline attribute (only applicable to iOS 10+)
   */
  static supported(mediaType?: Uiza.MediaType, provider?: Uiza.Provider, playsInline?: boolean): Uiza.Support;

  constructor(targets: NodeList | HTMLElement | HTMLElement[] | string, options?: Uiza.Options);

  /**
   * Indicates if the current player is HTML5.
   */
  readonly isHTML5: boolean;

  /**
   * Indicates if the current player is an embedded player.
   */
  readonly isEmbed: boolean;

  /**
   * Indicates if the current player is playing.
   */
  readonly playing: boolean;

  /**
   * Indicates if the current player is paused.
   */
  readonly paused: boolean;

  /**
   * Indicates if the current player is stopped.
   */
  readonly stopped: boolean;

  /**
   * Indicates if the current player has finished playback.
   */
  readonly ended: boolean;

  /**
   * Returns a float between 0 and 1 indicating how much of the media is buffered
   */
  readonly buffered: number;

  /**
   * Gets or sets the currentTime for the player. The setter accepts a float in seconds.
   */
  currentTime: number;

  /**
   * Indicates if the current player is seeking.
   */
  readonly seeking: boolean;

  /**
   * Returns the duration for the current media.
   */
  readonly duration: number;

  /**
   * Gets or sets the volume for the player. The setter accepts a float between 0 and 1.
   */
  volume: number;

  /**
   * Gets or sets the muted state of the player. The setter accepts a boolean.
   */
  muted: boolean;

  /**
   * Indicates if the current media has an audio track.
   */
  readonly hasAudio: boolean;

  /**
   * Gets or sets the speed for the player. The setter accepts a value in the options specified in your config. Generally the minimum should be 0.5.
   */
  speed: number;

  /**
   * Gets or sets the quality for the player. The setter accepts a value from the options specified in your config.
   */
  quality: number;

  /**
   * Gets or sets the current loop state of the player.
   */
  loop: boolean;

  /**
   * Gets or sets the current source for the player.
   */
  source: Uiza.SourceInfo;

  /**
   * Gets or sets the current poster image URL for the player.
   */
  poster: string;

  /**
   * Gets or sets the autoplay state of the player.
   */
  autoplay: boolean;

  /**
   * Gets or sets the caption track by index. 1 means the track is missing or captions is not active
   */
  currentTrack: number;

  /**
   * Gets or sets the preferred captions language for the player. The setter accepts an ISO twoletter language code. Support for the languages is dependent on the captions you include.
   * If your captions don't have any language data, or if you have multiple tracks with the same language, you may want to use currentTrack instead.
   */
  language: string;

  /**
   * Gets or sets the picture-in-picture state of the player. This currently only supported on Safari 10+ on MacOS Sierra+ and iOS 10+.
   */
  pip: boolean;

  /**
   * Gets or sets the aspect ratio for embedded players.
   */
  ratio?: string;

  /**
   * Returns the current video Provider
   */
  readonly provider: 'html5' | 'hlsjs';

  /**
   * Returns the native API for Vimeo or Youtube players
   */
  readonly embed?: any;

  readonly fullscreen: Uiza.FullscreenControl;

  /**
   * Start playback.
   * For HTML5 players, play() will return a Promise in some browsers - WebKit and Mozilla according to MDN at time of writing.
   */
  play(): Promise<void> | void;

  /**
   * Pause playback.
   */
  pause(): void;

  /**
   * Toggle playback, if no parameters are passed, it will toggle based on current status.
   */
  togglePlay(toggle?: boolean): boolean;

  /**
   * Stop playback and reset to start.
   */
  stop(): void;

  /**
   * Restart playback.
   */
  restart(): void;

  /**
   * Rewind playback by the specified seek time. If no parameter is passed, the default seek time will be used.
   */
  rewind(seekTime?: number): void;

  /**
   * Fast forward by the specified seek time. If no parameter is passed, the default seek time will be used.
   */
  forward(seekTime?: number): void;

  /**
   * Increase volume by the specified step. If no parameter is passed, the default step will be used.
   */
  increaseVolume(step?: number): void;

  /**
   * Increase volume by the specified step. If no parameter is passed, the default step will be used.
   */
  decreaseVolume(step?: number): void;

  /**
   * Toggle captions display. If no parameter is passed, it will toggle based on current status.
   */
  toggleCaptions(toggle?: boolean): void;

  /**
   * Trigger the airplay dialog on supported devices.
   */
  airplay(): void;

  /**
   * Toggle the controls (video only). Takes optional truthy value to force it on/off.
   */
  toggleControls(toggle: boolean): void;

  /**
   * Add an event listener for the specified event.
   */
  on(
    event: Uiza.StandardEvent | Uiza.Html5Event | Uiza.YoutubeEvent,
    callback: (this: this, event: Uiza.UizaEvent) => void,
  ): void;

  /**
   * Add an event listener for the specified event once.
   */
  once(
    event: Uiza.StandardEvent | Uiza.Html5Event | Uiza.YoutubeEvent,
    callback: (this: this, event: Uiza.UizaEvent) => void,
  ): void;

  /**
   * Remove an event listener for the specified event.
   */
  off(
    event: Uiza.StandardEvent | Uiza.Html5Event | Uiza.YoutubeEvent,
    callback: (this: this, event: Uiza.UizaEvent) => void,
  ): void;

  /**
   * Check support for a mime type.
   */
  supports(type: string): boolean;

  /**
   * Destroy lib instance
   */
  destroy(): void;
}

declare namespace Uiza {
  type MediaType = 'audio' | 'video';
  type Provider = 'html5' | 'hlsjs';
  type StandardEvent =
    | 'progress'
    | 'playing'
    | 'play'
    | 'pause'
    | 'timeupdate'
    | 'volumechange'
    | 'seeking'
    | 'seeked'
    | 'ratechange'
    | 'ended'
    | 'enterfullscreen'
    | 'exitfullscreen'
    | 'captionsenabled'
    | 'captionsdisabled'
    | 'languagechange'
    | 'controlshidden'
    | 'controlsshown'
    | 'ready';
  type Html5Event =
    | 'loadstart'
    | 'loadeddata'
    | 'loadedmetadata'
    | 'canplay'
    | 'canplaythrough'
    | 'stalled'
    | 'waiting'
    | 'emptied'
    | 'cuechange'
    | 'error';

  interface FullscreenControl {
    /**
     * Indicates if the current player is in fullscreen mode.
     */
    readonly active: boolean;

    /**
     * Indicates if the current player has fullscreen enabled.
     */
    readonly enabled: boolean;

    /**
     * Enter fullscreen. If fullscreen is not supported, a fallback ""full window/viewport"" is used instead.
     */
    enter(): void;

    /**
     * Exit fullscreen.
     */
    exit(): void;

    /**
     * Toggle fullscreen.
     */
    toggle(): void;
  }

  interface Options {
    /**
     * Completely disable Uiza. This would allow you to do a User Agent check or similar to programmatically enable or disable Uiza for a certain UA. Example below.
     */
    enabled?: boolean;

    /**
     * Display debugging information in the console
     */
    debug?: boolean;

    /**
     * If a function is passed, it is assumed your method will return either an element or HTML string for the controls. Three arguments will be passed to your function;
     * id (the unique id for the player), seektime (the seektime step in seconds), and title (the media title). See CONTROLS.md for more info on how the html needs to be structured.
     * Defaults to ['play-large', 'play', 'progress', 'current-time', 'mute', 'volume', 'captions', 'settings', 'pip', 'airplay', 'fullscreen']
     */
    controls?: string[] | ((id: string, seektime: number, title: string) => unknown) | Element;

    /**
     * If you're using the default controls are used then you can specify which settings to show in the menu
     * Defaults to ['captions', 'quality', 'speed', 'loop']
     */
    settings?: string[];

    /**
     * Used for internationalization (i18n) of the text within the UI.
     */
    i18n?: any;

    /**
     * Load the SVG sprite specified as the iconUrl option (if a URL). If false, it is assumed you are handling sprite loading yourself.
     */
    loadSprite?: boolean;

    /**
     * Specify a URL or path to the SVG sprite. See the SVG section for more info.
     */
    iconUrl?: string;

    /**
     * Specify the id prefix for the icons used in the default controls (e.g. Uiza-play would be Uiza).
     * This is to prevent clashes if you're using your own SVG sprite but with the default controls.
     * Most people can ignore this option.
     */
    iconPrefix?: string;

    /**
     * Specify a URL or path to a blank video file used to properly cancel network requests.
     */
    blankUrl?: string;

    /**
     * Autoplay the media on load. This is generally advised against on UX grounds. It is also disabled by default in some browsers.
     * If the autoplay attribute is present on a <video> or <audio> element, this will be automatically set to true.
     */
    autoplay?: boolean;

    /**
     * Only allow one player playing at once.
     */
    autopause?: boolean;

    /**
     * The time, in seconds, to seek when a user hits fast forward or rewind.
     */
    seekTime?: number;

    /**
     * A number, between 0 and 1, representing the initial volume of the player.
     */
    volume?: number;

    /**
     * Whether to start playback muted. If the muted attribute is present on a <video> or <audio> element, this will be automatically set to true.
     */
    muted?: boolean;

    /**
     * Click (or tap) of the video container will toggle play/pause.
     */
    clickToPlay?: boolean;

    /**
     * Disable right click menu on video to help as very primitive obfuscation to prevent downloads of content.
     */
    disableContextMenu?: boolean;

    /**
     * Hide video controls automatically after 2s of no mouse or focus movement, on control element blur (tab out), on playback start or entering fullscreen.
     * As soon as the mouse is moved, a control element is focused or playback is paused, the controls reappear instantly.
     */
    hideControls?: boolean;

    /**
     * Reset the playback to the start once playback is complete.
     */
    resetOnEnd?: boolean;

    /**
     * Enable keyboard shortcuts for focused players only or globally
     */
    keyboard?: KeyboardOptions;

    /**
     * controls: Display control labels as tooltips on :hover & :focus (by default, the labels are screen reader only).
     * seek: Display a seek tooltip to indicate on click where the media would seek to.
     */
    tooltips?: TooltipOptions;

    /**
     * Specify a custom duration for media.
     */
    duration?: number;

    /**
     * Displays the duration of the media on the metadataloaded event (on startup) in the current time display.
     * This will only work if the preload attribute is not set to none (or is not set at all) and you choose not to display the duration (see controls option).
     */
    displayDuration?: boolean;

    /**
     * Display the current time as a countdown rather than an incremental counter.
     */
    invertTime?: boolean;

    /**
     * Allow users to click to toggle the above.
     */
    toggleInvert?: boolean;

    /**
     * Allows binding of event listeners to the controls before the default handlers. See the defaults.js for available listeners.
     * If your handler prevents default on the event (event.preventDefault()), the default handler will not fire.
     */
    listeners?: { [key: string]: (error: UizaEvent) => void };

    /**
     * active: Toggles if captions should be active by default. language: Sets the default language to load (if available). 'auto' uses the browser language.
     * update: Listen to changes to tracks and update menu. This is needed for some streaming libraries, but can result in unselectable language options).
     */
    captions?: CaptionOptions;

    /**
     * enabled: Toggles whether fullscreen should be enabled. fallback: Allow fallback to a full-window solution.
     * iosNative: whether to use native iOS fullscreen when entering fullscreen (no custom controls)
     */
    fullscreen?: FullScreenOptions;

    /**
     * The aspect ratio you want to use for embedded players.
     */
    ratio?: string;

    /**
     * enabled: Allow use of local storage to store user settings. key: The key name to use.
     */
    storage?: StorageOptions;

    /**
     * selected: The default speed for playback. options: The speed options to display in the UI. YouTube and Vimeo will ignore any options outside of the 0.5-2 range, so options outside of this range will be hidden automatically.
     */
    speed?: SpeedOptions;

    /**
     * Currently only supported by YouTube. default is the default quality level, determined by YouTube. options are the options to display.
     */
    quality?: QualityOptions;

    /**
     * active: Whether to loop the current video. If the loop attribute is present on a <video> or <audio> element,
     * this will be automatically set to true This is an object to support future functionality.
     */
    loop?: LoopOptions;

    /**
     * Preview Thumbnails Options.
     */
    previewThumbnails?: PreviewThumbnailsOptions;
  }

  interface QualityOptions {
    default: number;
    options: number[];
  }

  interface LoopOptions {
    active: boolean;
  }

  interface AdOptions {
    enabled: boolean;
    publisherId: string;
  }

  interface SpeedOptions {
    selected: number;
    options: number[];
  }

  interface KeyboardOptions {
    focused?: boolean;
    global?: boolean;
  }

  interface TooltipOptions {
    controls?: boolean;
    seek?: boolean;
  }

  interface FullScreenOptions {
    enabled?: boolean;
    fallback?: boolean;
    allowAudio?: boolean;
    iosNative?: boolean;
  }

  interface CaptionOptions {
    active?: boolean;
    language?: string;
    update?: boolean;
  }

  interface StorageOptions {
    enabled?: boolean;
    key?: string;
  }

  interface PreviewThumbnailsOptions {
    enabled?: boolean;
    src?: string;
  }

  interface SourceInfo {
    /**
     * Note: YouTube and Vimeo are currently not supported as audio sources.
     */
    type: MediaType;

    /**
     * Title of the new media. Used for the aria-label attribute on the play button, and outer container. YouTube and Vimeo are populated automatically.
     */
    title?: string;

    /**
     * This is an array of sources. For HTML5 media, the properties of this object are mapped directly to HTML attributes so more can be added to the object if required.
     */
    sources: Source[];

    /**
     * The URL for the poster image (HTML5 video only).
     */
    poster?: string;

    /**
     * An array of track objects. Each element in the array is mapped directly to a track element and any keys mapped directly to HTML attributes so as in the example above,
     * it will render as <track kind="captions" label="English" srclang="en" src="https://cdn.selz.com/Uiza/1.0/example_captions_en.vtt" default> and similar for the French version.
     * Booleans are converted to HTML5 value-less attributes.
     */
    tracks?: Track[];
  }

  interface Source {
    /**
     * The URL of the media file (or YouTube/Vimeo URL).
     */
    src: string;
    /**
     * The MIME type of the media file (if HTML5).
     */
    type?: string;
    provider?: Provider;
    size?: number;
  }

  type TrackKind = 'subtitles' | 'captions' | 'descriptions' | 'chapters' | 'metadata';
  interface Track {
    /**
     * Indicates how the text track is meant to be used
     */
    kind: TrackKind;
    /**
     * Indicates a user-readable title for the track
     */
    label: string;
    /**
     * The language of the track text data. It must be a valid BCP 47 language tag. If the kind attribute is set to subtitles, then srclang must be defined.
     */
    srcLang?: string;
    /**
     * The URL of the track (.vtt file).
     */
    src: string;

    default?: boolean;
  }

  interface UizaEvent extends CustomEvent {
    readonly detail: { readonly Uiza: Uiza };
  }

  interface Support {
    api: boolean;
    ui: boolean;
  }
}
